flake.nix

```diff
@@ -268,7 +268,28 @@
           # Build *just* the cargo dependencies, so we can reuse all of that work (e.g. via cachix) when running in CI
           cargoArtifacts = craneLib.buildDepsOnly {
             inherit pname src;
-            cargoExtraArgs = "${cargoBuildExtraArgs} --workspace --all-features";
+            cargoExtraArgs = "${cargoBuildExtraArgs} --all-features";
+            # If we build all the packages at once, feature unification takes
+            # over and we get libraries with different sets of features than
+            # we would get building them separately. Meaning that when we
+            # later build them separately, it won't hit the cache. So instead,
+            # we need to build each package separately when we are collecting
+            # dependencies.
+            cargoBuildCommand = "cargoWorkspace build";
+            cargoTestCommand = "cargoWorkspace test";
+            cargoCheckCommand = "cargoWorkspace check";
+            preBuild = ''
+              cargoWorkspace() {
+                command=$(shift)
+                for packageDir in $(${pkgs.yq}/bin/tomlq -r '.workspace.members[]' Cargo.toml); do
+                  (
+                    cd $packageDir
+                    pwd
+                    cargoWithProfile $command "$@"
```

Comment:
Where's `cargoWithProfile` defined?

Reply:
somewhere in `craneLib`. it's the default value of `cargo{Build,Test,Check}Command`